File: apps/shared/src/wire/event.ts
// SessionEvent = append-only log row. Coord's `events` table is the
// source of truth; Session projections = fold(events). Browser folds
// the same events from the SSE stream. R0.3, R3.1.

import { z } from "zod";
import { Session, SessionKind } from "./session.ts";
import { ChannelId, SessionId, WorkerFp, WorkspaceId, TraceId } from "./brand.ts";

const Base = z.object({
  ts: z.number().int().positive(),
  trace_id: TraceId.optional(),
});

export const SessionEvent = z.discriminatedUnion("kind", [
  Base.extend({
    kind: z.literal("opened"),
    session_id: SessionId,
    worker_fp: WorkerFp,
    channel: ChannelId,
    session_kind: SessionKind,
    cwd: z.string(),
  }),
  Base.extend({
    kind: z.literal("closed"),
    session_id: SessionId,
    exit_code: z.number().int().nullable(),
  }),
  Base.extend({
    kind: z.literal("attached"),
    session_id: SessionId,
  }),
  Base.extend({
    kind: z.literal("detached"),
    session_id: SessionId,
  }),
  Base.extend({
    kind: z.literal("cwd"),
    session_id: SessionId,
    cwd: z.string(),
  }),
  Base.extend({
    kind: z.literal("workspace_assigned"),
    session_id: SessionId,
    workspace_id: WorkspaceId.nullable(),
  }),
  Base.extend({
    // Worker re-announces all live sessions on coord reconnect.
    // Coord reconciles: any DB session for this worker NOT in the
    // snapshot gets a synthetic `closed` event appended.
    kind: z.literal("snapshot"),
    worker_fp: WorkerFp,
    sessions: z.array(Session),
  }),
  Base.extend({
    // Worker rebooted; the keeper PTY for this session died. Worker
    // spawned a fresh PTY at the same cwd and re-bound it to the same
    // session_id. Sidebar row stays in place.
    kind: z.literal("respawned"),
    session_id: SessionId,
    new_channel: ChannelId,
  }),
  Base.extend({
    // User renamed the session from the sidebar. custom_title="" clears the
    // override (revert to auto title). Sticky: auto-title events never touch
    // custom_title, so a rename survives OSC-title churn.
    kind: z.literal("renamed"),
    session_id: SessionId,
    custom_title: z.string(),
  }),
  Base.extend({
    // Worker resolved / re-resolved the git branch of the session's cwd.
    // branch=null → folder isn't a git repo. Feeds the cell-grid folder-row
    // subtitle (FolderList.tsx). Emitted on spawn + on .git/HEAD change.
    kind: z.literal("git"),
    session_id: SessionId,
    branch: z.string().nullable(),
    // GitHub owner/repo — present only when resolved; absent = leave alone.
    remote: z.string().optional(),
  }),
  Base.extend({
    // Worker-resolved GitHub PR status for the session's branch (via
    // `gh pr list --head <branch>`). number=null → no open PR for the branch.
    // Feeds the #123 ✓ folder-row badge (FolderList.tsx). Emitted on
    // spawn + branch-change + a 90s poll, deduped to changes only.
    kind: z.literal("pr"),
    session_id: SessionId,
    number: z.number().int().nullable(),
    state: z.enum(["open", "merged", "closed", "draft"]).nullable(),
    checks: z.enum(["passing", "failing", "pending", "none"]).nullable(),
    url: z.string().nullable(),
  }),
  Base.extend({
    // Worker-detected LISTEN ports of the session's process tree (ps+lsof).
    // Feeds the :5174 folder-row chips (FolderList.tsx). Emitted on spawn +
    // a 90s poll, deduped to changes. Empty = nothing listening.
    kind: z.literal("ports"),
    session_id: SessionId,
    ports: z.array(z.number().int()),
  }),
]);
export type SessionEvent = z.infer<typeof SessionEvent>;

// Fold one event into a session map. Pure function — replay is
// deterministic so long as the input order is stable.
export function foldEvent(
  prev: Map<string, Session>,
  e: SessionEvent,
): Map<string, Session> {
  const next = new Map(prev);
  switch (e.kind) {
    case "opened":
      next.set(e.session_id, {
        id: e.session_id,
        worker_fp: e.worker_fp,
        channel: e.channel,
        kind: e.session_kind,
        cwd: e.cwd,
        // Immutable spawn folder — captured once here from the opened cwd. The
        // `cwd` event below updates `cwd` only, never this. Backs the /t/ URL.
        spawn_cwd: e.cwd,
        workspace_id: null,
        status: "open",
        created_at: e.ts,
        closed_at: null,
        custom_title: null,
      });
      return next;
    case "closed": {
      // The terminal's process actually exited (worker emits `closed` only on
      // real PTY exit) or coord confirmed it's gone from the worker's
      // authoritative live snapshot (ghost). DELETE the row — no "closed"
      // limbo. A LIVE terminal is always in the worker snapshot, so it is
      // never seen as closed and never reaches this branch. `closed` is the
      // ONLY deletion trigger.
      if (!prev.has(e.session_id)) return prev;
      next.delete(e.session_id);
      return next;
    }
    case "attached":
    case "detached":
      return prev; // pure liveness; no projection change yet
    case "cwd": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      next.set(e.session_id, { ...s, cwd: e.cwd });
      return next;
    }
    case "workspace_assigned": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      next.set(e.session_id, { ...s, workspace_id: e.workspace_id });
      return next;
    }
    case "snapshot": {
      // Sync this worker's ANNOUNCED sessions into the projection (upsert). A
      // session of this worker that is ABSENT from the snapshot is NOT deleted —
      // it persists as an offline "breadcrumb" row: a worker restart kills the
      // PTY but the row survives so the sidebar still shows where you were
      // working. A reconnect respawn re-binds it; an explicit `closed` (real PTY
      // exit) or the user's ✕ removes it. Other workers' sessions are untouched.
      // (This mirrors coord's SQL snapshot reconcile in event-log.ts — both stop
      // pruning ghosts, keeping the SPA + coord projections in agreement.)
      for (const s of e.sessions) {
        const before = prev.get(s.id);
        // These fields are coordinator/DB-owned or immutable creation facts.
        // A reconnect snapshot refreshes the live worker state without
        // rewriting history or user organization.
        next.set(s.id, before
          ? {
              ...s,
              created_at: before.created_at,
              workspace_id: before.workspace_id,
              custom_title: before.custom_title,
              spawn_cwd: before.spawn_cwd,
            }
          : s);
      }
      return next;
    }
    case "respawned": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      // New keeper channel. Status is forced open in case a prior 'closed'
      // was projected before the respawn lands.
      next.set(e.session_id, {
        ...s,
        channel: e.new_channel,
        status: "open",
        closed_at: null,
      });
      return next;
    }
    case "renamed": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      // "" clears the override → revert to auto title. Auto-title events
      // (OSC/cwd) never write custom_title, so this is sticky.
      next.set(e.session_id, { ...s, custom_title: e.custom_title || null });
      return next;
    }
    case "git": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      // branch always present; remote only when the worker resolved a github
      // origin (absent → preserve the prior git_remote).
      next.set(e.session_id, e.remote !== undefined
        ? { ...s, git_branch: e.branch, git_remote: e.remote }
        : { ...s, git_branch: e.branch });
      return next;
    }
    case "pr": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      next.set(e.session_id, {
        ...s,
        pr_number: e.number,
        pr_state: e.state,
        pr_checks: e.checks,
        pr_url: e.url,
      });
      return next;
    }
    case "ports": {
      const s = prev.get(e.session_id);
      if (!s) return prev;
      next.set(e.session_id, { ...s, ports: e.ports });
      return next;
    }
  }
}

export function foldAll(events: SessionEvent[]): Map<string, Session> {
  let acc = new Map<string, Session>();
  for (const e of events) acc = foldEvent(acc, e);
  return acc;
}

File: crates/roost-protocol/src/wire/event.rs
//! The canonical fold of the append-only session event log. There is exactly
//! one fold in Roost and it is here.
//!
//! Public session state is the deterministic fold of an ordered event log,
//! shared by the coordinator's projector, the browser's store, and the worker's
//! own reconciliation. Replay is deterministic for a stable input order, which
//! is what makes a projection rebuildable from the log at all. The union being
//! folded is in `variants`.
//!
//! `fold_event` takes its input by reference and returns a new map, so a caller
//! may hold a projection across a fold and know it was not touched. A no-op
//! variant, and any variant naming a session the projection does not hold, hands
//! back the same contents.
mod variants;

use std::collections::BTreeMap;

pub use self::variants::SessionEvent;
use crate::wire::brand::SessionId;
use crate::wire::session::{Session, SessionStatus};

/// The projection the fold produces, keyed by session id.
pub type SessionMap = BTreeMap<SessionId, Session>;

/// Fold one event into a session map. The input is never mutated: a no-op
/// variant and a variant naming a session the projection does not hold both
/// hand back the same contents.
pub fn fold_event(prev: &SessionMap, event: &SessionEvent) -> SessionMap {
    match event {
        // Private and viewer-local variants are explicit no-ops: they are
        // durable and ordered, but they are not public session state.
        SessionEvent::Attached { .. }
        | SessionEvent::Detached { .. }
        | SessionEvent::AgentReference { .. } => prev.clone(),

        // The only arm that ignores the previous state. It captures the spawn
        // folder once, and leaves the resolved git, pull-request, and port
        // fields absent rather than null: nothing has looked yet.
        SessionEvent::Opened {
            session_id,
            worker_fp,
            channel,
            session_kind,
            cwd,
            ts,
            ..
        } => {
            let mut next = prev.clone();
            next.insert(
                session_id.clone(),
                Session {
                    id: session_id.clone(),
                    worker_fp: worker_fp.clone(),
                    channel: *channel,
                    kind: *session_kind,
                    cwd: cwd.clone(),
                    spawn_cwd: Some(cwd.clone()),
                    workspace_id: None,
                    status: SessionStatus::Open,
                    created_at: *ts,
                    closed_at: None,
                    custom_title: None,
                    git_branch: None,
                    git_remote: None,
                    pr_number: None,
                    pr_state: None,
                    pr_checks: None,
                    pr_url: None,
                    ports: None,
                },
            );
            next
        }

        // The terminal's process actually exited, or the coordinator confirmed
        // the session is gone from the worker's authoritative live snapshot.
        // This is the only deletion trigger, and it is a no-op for a session the
        // projection does not hold.
        SessionEvent::Closed { session_id, .. } => {
            let mut next = prev.clone();
            next.remove(session_id);
            next
        }

        SessionEvent::Cwd {
            session_id, cwd, ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                // Only the live folder drifts. `spawn_cwd` is the creation
                // fact behind the stable `/t/` URL and never moves.
                session.cwd = cwd.clone();
            }
            next
        }

        SessionEvent::WorkspaceAssigned {
            session_id,
            workspace_id,
            ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.workspace_id = workspace_id.clone();
            }
            next
        }

        // Sync this worker's announced sessions into the projection. A session
        // of this worker that is absent from the snapshot is NOT deleted: it
        // persists as an offline breadcrumb, because a worker restart kills the
        // PTY but the row is where the user was working. Other workers' rows
        // are untouched. Exactly four fields are carried over from the previous
        // row — the coordinator- or user-owned ones the worker does not track.
        SessionEvent::Snapshot { sessions, .. } => {
            let mut next = prev.clone();
            for announced in sessions {
                let mut row = announced.clone();
                if let Some(before) = prev.get(&row.id) {
                    row.created_at = before.created_at;
                    row.workspace_id = before.workspace_id.clone();
                    row.custom_title = before.custom_title.clone();
                    row.spawn_cwd = before.spawn_cwd.clone();
                }
                next.insert(announced.id.clone(), row);
            }
            next
        }

        SessionEvent::Respawned {
            session_id,
            new_channel,
            ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.channel = *new_channel;
                // Forced open, in case a `closed` was projected before the
                // respawn landed.
                session.status = SessionStatus::Open;
                session.closed_at = None;
            }
            next
        }

        SessionEvent::Renamed {
            session_id,
            custom_title,
            ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.custom_title = if custom_title.is_empty() {
                    None
                } else {
                    Some(custom_title.clone())
                };
            }
            next
        }

        SessionEvent::Git {
            session_id,
            branch,
            remote,
            ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.git_branch = branch.clone();
                if let Some(remote) = remote {
                    session.git_remote = Some(Some(remote.clone()));
                }
            }
            next
        }

        SessionEvent::Pr {
            session_id,
            number,
            state,
            checks,
            url,
            ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.pr_number = *number;
                session.pr_state = *state;
                session.pr_checks = *checks;
                session.pr_url = url.clone();
            }
            next
        }

        SessionEvent::Ports {
            session_id, ports, ..
        } => {
            let mut next = prev.clone();
            if let Some(session) = next.get_mut(session_id) {
                session.ports = Some(ports.clone());
            }
            next
        }
    }
}

/// Fold an ordered log from nothing. Replay is deterministic, so rebuilding a
/// projection from the log reproduces it exactly.
pub fn fold_all(events: &[SessionEvent]) -> SessionMap {
    let mut projection = SessionMap::new();
    for event in events {
        projection = fold_event(&projection, event);
    }
    projection
}
